### PR TITLE
11/refactor

### LINKED ISSUE #5: First recycling action does not work
It seems that the first recycling event does create the recyclebin item but cannot recycle the item right away. This might come from the fact the newly recyclebin item is not yet available when running the second task? This bug might be resolved when refactoring the backend task management.

### LINKED ISSUE #10: Restore recycled items
This is currently very difficult to implement, as there is no clean way to save the original path.
Probably we should go with #2 where this kind of info can be saved.

### LINKED ISSUE #11: Refactor given new task manager

### PR DESCRIPTION
**Pull request by pyphilia**
This PR completely refactor the plugin. It is not perfect, as it needs to check everytime the status of the item before proceeding.

#2 proposed a table to stock items, the problem with this idea is the `ON DELETE CASCADE` of memberships. Thus deleting/moving an item to another table will delete the corresponding memberships. This is a problem as restoring should restore the item in its entirety. 

One pro of this new implementation is that multiple members could have the same deleted item in its trash, which might be interesting for shared items.

close #5 
close #10 
close #11

### CODE REVIEW

File: src/tasks/delete-recycled-item-task.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    input?: DeleteRecycledItemTaskInput,
+  ) {
+    super(member, recycleItemService);
+    this.input = input;
+  }
+
+  async run(handler: DatabaseTransactionHandler): Promise<void> {
+    this.status = 'RUNNING';
+
+    // get member's "own" recycled items (member is admin and is in recycled items table)
```
Comment (Julien-Torrent) [inline src/tasks/delete-recycled-item-task.ts:29]: Comment doesn't match with code

File: src/plugin.ts [hunk 2]
```diff
@@ -1,193 +1,199 @@
...
 
-    // remove all memberships from item (and its subtree)
-    const cleanItemMemeberships =
-      itemMembershipTaskManager.createDeleteAllOnAndBelowItemTask(member, itemId);
-    await runner.runSingle(cleanItemMemeberships, log);
+    // create entry in table
+    const t2 = recycledItemTaskManager.createCreateTask(member, {});
+    t2.getInput = () => t1[0].result;
+
+    await runner.runSingleSequence([...t1, t3, t4, t2], log);
```
Comment (Julien-Torrent) [inline src/plugin.ts:181]: Tasks name are not in order

File: src/tasks/is-item-deleted-task.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    this.input = input;
+  }
+
+  async run(handler: DatabaseTransactionHandler): Promise<void> {
+    this.status = 'RUNNING';
+
+    const { item, validate } = this.input;
+    const isDeleted = await this.recycleItemService.isDeleted(item.path, handler);
+
+    // if item should be validate, throw if it doesn't satisfy validate input
```
Comment (Julien-Torrent) [inline src/tasks/is-item-deleted-task.ts:31]: Does this means if I set validate to true, the task returns the item only if it's deleted ?

File: db-schema.sql [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+-- get children of deleted item
+    -- get item normal? <- front can infer recycled from path given id
+
+-- get recycle bin items
+    -- get children of recyclebin item <- does not work bc regular call doesn't check memberships
+    ---> join recycle table x admin memberships <<<- allow for multiple people to have the same item
+    
+-- XX move whole item to recycle
+--     - move memberships??
```
Comment (Julien-Torrent) [inline db-schema.sql:67]: Delete unused comments

File: src/tasks/is-item-deleted-task.ts (hunk 3 above)
Comment (pyphilia) [reply src/tasks/is-item-deleted-task.ts:31]: Yes exactly.

**Pull request merged**